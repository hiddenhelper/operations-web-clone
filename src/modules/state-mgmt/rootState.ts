import { History } from 'history';

import { IState as IAuthState } from './auth/state';
import { IState as IGeneralState } from './general/state';
import { IState as IClientState } from './client/state';
import { IState as IProjectState } from './project/state';
import { IState as IUserState } from './user/state';
import { IState as IAccessControlSystemState } from './access-control-system/state';
import { IState as IBadgePrintingSystemState } from './badge-printer-system/state';
import { IState as IWorkerState } from './worker/state';
import { IState as ICertificationState } from './certification/state';
import { IState as IFileState } from './file/state';
import { IState as IBadgeState } from './badge/state';
import { IState as IStatisticsState } from './statistics/state';
import { IState as ITrainingState } from './training/state';
import { IState as IPaymentState } from './payment/state';
import { IState as IInvoiceState } from './invoice/state';
import { ApiService } from '../services/ApiService';
import { AuthService } from '../services/AuthService';
import { PrinterService } from '../services/PrinterService';
import { FileService } from '../services/FileService';
import { Logger } from '../services/Logger';
import { IProcoreState } from './procore/state';
import { IState as IProjectNewState } from './project-new/state';

export interface IAction {
  type: string;
  payload: any;
}

export interface IRootState {
  auth: IAuthState;
  user: IUserState;
  client: IClientState;
  project: IProjectState;
  accessControlSystem: IAccessControlSystemState;
  badgePrinterSystem: IBadgePrintingSystemState;
  worker: IWorkerState;
  certification: ICertificationState;
  training: ITrainingState;
  invoice: IInvoiceState;
  file: IFileState;
  badge: IBadgeState;
  statistics: IStatisticsState;
  payment: IPaymentState;
  general: IGeneralState;
  router?: { location: { pathname: string; search: string; hard: string; key: string }; action: string };
  procore: IProcoreState;
  projectNew: IProjectNewState;
}

export interface IEpicDependencies {
  history: History;
  apiService: ApiService;
  authService: AuthService;
  printerService: PrinterService;
  fileService: FileService;
  logger: typeof Logger;
}
