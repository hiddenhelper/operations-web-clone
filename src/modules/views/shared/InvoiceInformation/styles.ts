import { createStyles, makeStyles } from '@material-ui/core/styles';
import { toREM } from '../../../../utils/generalUtils';
import { STYLE } from '../../../../constants';

export const useStyles = makeStyles(theme =>
  createStyles({
    textColor: {
      color: STYLE.COLOR.SECONDARY_DARKER,
    },
    invoiceStatusChip: {
      marginLeft: '10px',
    },
    invoiceStatusWrapper: {
      position: 'relative',
      top: '2px',
    },
    invoiceTitleDate: {
      fontWeight: 600,
      fontSize: `${toREM(15)}`,
      position: 'relative',
      top: '2px',
    },
    infoWrapper: {
      fontSize: `${toREM(17)}`,
    },
    infoAccent: {
      fontWeight: 600,
    },
    breakWord: {
      wordBreak: 'break-word',
    },
    infoDivider: {
      borderRight: `solid 1px ${STYLE.COLOR.LIGHT_GRAY3}`,
    },
    infoSecondItem: {
      paddingLeft: 20,
    },
    divider: {
      margin: '21px 0px',
      marginBottom: '30px !important',
      backgroundColor: STYLE.COLOR.DIVIDER,
    },
    subtitle: {
      fontSize: `${toREM(20)}`,
      fontWeight: 600,
    },
    serviceNameCell: {
      padding: '10px 0px 8px 16px',

      '& .MuiSelect-root': {
        width: '85%',
      },

      '& .MuiOutlinedInput-notchedOutline': {
        border: 0,
      },

      '& .MuiOutlinedInput-root': {
        backgroundColor: STYLE.COLOR.WHITE,
      },

      '& .MuiOutlinedInput-input': {
        backgroundColor: STYLE.COLOR.WHITE,
        padding: '10px 0px',
        fontWeight: 400,
      },

      '& .MuiSelect-root:focus': {
        backgroundColor: STYLE.COLOR.WHITE,
      },
    },
    secondarySubtitle: {
      fontSize: `${toREM(15)}`,
      fontWeight: 600,
      marginTop: 30,
      marginBottom: 10,
    },
  })
);
