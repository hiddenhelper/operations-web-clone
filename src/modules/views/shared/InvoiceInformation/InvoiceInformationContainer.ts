import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { IRootState } from '../../../state-mgmt/rootState';
import { invoiceState } from '../../../state-mgmt/invoice';

import { GENERAL } from '../../../../constants';
import InvoiceInformation from './InvoiceInformation';

export const mapStateToProps = (state: IRootState) => ({
  invoiceMap: state.invoice.invoiceMap,
  loading: state.general.loadingMap[GENERAL.LOADING_KEY.FETCH_INVOICE],
  downloadLoading: state.general.loadingMap[GENERAL.LOADING_KEY.DOWNLOAD_INVOICE],
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchInvoiceInformation: (id: string) => dispatch(invoiceState.actions.fetchInvoiceStart(id)),
  downloadInvoice: (id: string, name: string) => dispatch(invoiceState.actions.downloadStart(id, name)),
});

export default connect(mapStateToProps, mapDispatchToProps)(InvoiceInformation);
