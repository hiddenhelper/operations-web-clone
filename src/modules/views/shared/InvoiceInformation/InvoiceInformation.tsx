import React, { memo, useEffect, useCallback, useMemo } from 'react';
import moment from 'moment';

import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import DialogContent from '@material-ui/core/DialogContent';
import Grid from '@material-ui/core/Grid';
import Divider from '@material-ui/core/Divider';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import IconButton from '@material-ui/core/IconButton';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';

import Card from '../ResourceManagement/Card';
import Modal from '../Modal';
import Logo from '../Logo';
import StatusChip from '../StatusChip';
import InformationNote from '../InformationNote';
import Address from '../Address';

import { GeneralModel, InvoiceModel } from '../../../models';
import { LANG, CloseIcon, DownloadIcon } from '../../../../constants';
import {
  formatNumberWithCommas,
  getConditionalDefaultValue,
  getDefaultValue,
  getFormattedDate,
  getFormattedDecimalNumber,
} from '../../../../utils/generalUtils';
import { fullScreenModalGlobalStyles, invoiceFormGlobalStyles, tableGlobalStyles } from '../../../../assets/styles';
import { useStyles as statusChipStyles } from '../StatusChip/styles';
import { useStyles } from './styles';

export interface IInvoiceInformationProps {
  invoiceId: string;
  invoiceMap: GeneralModel.IEntityMap<InvoiceModel.IInvoice>;
  loading: GeneralModel.ILoadingStatus;
  downloadLoading: GeneralModel.ILoadingStatus;
  onClose: () => void;
  fetchInvoiceInformation: (id: string) => void;
  downloadInvoice: (id: string, name: string) => void;
}

const InvoiceInformation = ({ invoiceId, invoiceMap, loading, onClose, downloadInvoice, fetchInvoiceInformation }: IInvoiceInformationProps) => {
  const modalClasses = fullScreenModalGlobalStyles();
  const statusChipClasses = statusChipStyles();
  const tableGlobalClasses = tableGlobalStyles();
  const invoiceStyles = invoiceFormGlobalStyles();
  const classes = useStyles();

  const invoice = useMemo(() => invoiceMap[invoiceId], [invoiceMap, invoiceId]);

  const isDraft = useMemo(() => invoice?.status === InvoiceModel.InvoiceStatus.DRAFT, [invoice]);
  const isManual = useMemo(() => invoice?.type === InvoiceModel.InvoiceType.MANUAL, [invoice]);

  const invoiceStatusMap = useMemo(
    () => ({
      [InvoiceModel.InvoiceStatus.PENDING]: { label: 'Pending', class: statusChipClasses.pending },
      [InvoiceModel.InvoiceStatus.PAID]: { label: 'Paid', class: statusChipClasses.active },
      [InvoiceModel.InvoiceStatus.DECLINED]: { label: 'Declined', class: statusChipClasses.expired },
      [InvoiceModel.InvoiceStatus.PROCESSING]: { label: 'Processing', class: statusChipClasses.active },
      [InvoiceModel.InvoiceStatus.VOIDED]: { label: 'Voided', class: statusChipClasses.Finished },
    }),
    [statusChipClasses]
  );

  const invoiceList = useMemo(() => invoice?.items || [], [invoice]);

  const onDownloadInvoice = useCallback(() => {
    const invoiceCreationDate = moment(invoice.invoiceDate).format('YYYY-MM-DD');
    downloadInvoice(invoice.id, `${invoiceCreationDate} ${invoice.invoiceNumber}`);
  }, [invoice, downloadInvoice]);

  useEffect(() => {
    fetchInvoiceInformation(invoiceId);
  }, [invoiceId, fetchInvoiceInformation]);

  return (
    <Modal
      onClose={onClose}
      show={true}
      fullScreen={true}
      render={() => (
        <>
          {loading && loading.isLoading ? (
            <div className={modalClasses.loadingSkeleton}>Loading...</div>
          ) : (
            <>
              <div className={modalClasses.headerWrapper}>
                <div className={modalClasses.headerData}>
                  <Logo styleClass={modalClasses.logoWrapper} />
                </div>
                <div className={classes.invoiceWrapper}>
                  <div className={`${classes.titleCenter}`}>
                    <Typography className={`${modalClasses.title} ${modalClasses.smallDevice}`}>{getDefaultValue(invoice?.company?.name)}</Typography>
                    <Typography className={`${modalClasses.title} ${modalClasses.smallDevice}`}>{getDefaultValue(invoice?.project.name)}</Typography>
                    <Typography className={`${modalClasses.subTitle} ${modalClasses.smallDevice}`}>
                      {getFormattedDate(invoice?.invoiceDate, GeneralModel.DateFormat.MONTH_YEAR)}.{isManual && ' Manual Invoice.'}
                    </Typography>
                  </div>
                  {!isDraft && (
                    <div className={modalClasses.actionButtons}>
                      <IconButton disableRipple={true} onClick={onDownloadInvoice} data-testid="download-invoice">
                        <DownloadIcon />
                      </IconButton>
                    </div>
                  )}
                  <div>
                    <Button disableRipple={true} className={modalClasses.closeButton} data-testid="close-invoice-info" onClick={onClose}>
                      <CloseIcon />
                    </Button>
                  </div>
                </div>
              </div>
              <DialogContent className={`${modalClasses.cardDividerMargin} ${classes.dialogContent}`}>
                <Card
                  title={`Invoice #${invoice?.invoiceNumber}`}
                  actionStyleClass={classes.invoiceStatusWrapper}
                  secondaryAction={
                    <>
                      <span className={classes.invoiceTitleDate}>{getFormattedDate(invoice?.invoiceDate, GeneralModel.DateFormat.DATE)}</span>
                      <StatusChip
                        styleClasses={`${invoiceStatusMap[invoice?.status]?.class} ${classes.invoiceStatusChip}`}
                        label={invoiceStatusMap[invoice?.status]?.label}
                      />
                      <br />
                      <span className={`${classes.infoWrapper} ${classes.textColor}`}>Project: {getDefaultValue(invoice?.project.name)}</span>
                    </>
                  }
                >
                  <Grid container={true} className={`${classes.infoWrapper} ${classes.textColor}`}>
                    <Grid item={true} xl={1} lg={1} className={classes.infoAccent}>
                      From
                    </Grid>
                    <Grid item={true} xl={5} lg={5} className={`${classes.infoDivider}`}>
                      <div className={classes.infoAccent}>Field Control Analytics</div>
                      <div>
                        Suite 300, <span style={{ fontWeight: 'bold' }}>8350 North Central Expressway</span>
                      </div>
                      <div>Dallas, Texas, 75206, USA</div>
                    </Grid>
                    <Grid item={true} xl={1} lg={1} className={`${classes.infoAccent} ${classes.infoSecondItem}`}>
                      To
                    </Grid>
                    <Grid item={true} xl={5} lg={5} className={`${classes.infoSecondItem}`}>
                      <div className={classes.infoAccent}>{invoice?.company?.name}</div>
                      <Address address={invoice?.company?.billingAddress} bigFont={true} />
                      <div className={classes.breakWord}>
                        {`${invoice?.company?.user?.firstName} ${invoice?.company?.user?.lastName}`}: {invoice?.company?.user?.email}
                      </div>
                    </Grid>
                  </Grid>
                  <Divider className={classes.divider} />
                  <Typography className={`${classes.subtitle} ${classes.textColor}`}>Invoice Detail</Typography>
                  <Grid item={true} xl={12} lg={12} className={`${tableGlobalClasses.tableWrapper}`}>
                    <Table aria-label="invoice-services">
                      <TableHead>
                        <TableRow>
                          <TableCell>Date</TableCell>
                          <TableCell>Service</TableCell>
                          <TableCell>Detail</TableCell>
                          <TableCell>Amount</TableCell>
                        </TableRow>
                      </TableHead>
                      <TableBody>
                        {invoiceList.map((item, index) => (
                          <TableRow key={index}>
                            <TableCell className={classes.serviceNameCell}>{getFormattedDate(invoice.invoiceDate, GeneralModel.DateFormat.DATE)}</TableCell>
                            <TableCell className={classes.serviceNameCell}>{item.service?.name}</TableCell>
                            <TableCell className={classes.serviceNameCell}>{item.detail}</TableCell>
                            <TableCell className={`${classes.serviceNameCell} ${classes.infoAccent}`}>
                              $ {formatNumberWithCommas(getFormattedDecimalNumber(item.amount))}
                            </TableCell>
                          </TableRow>
                        ))}
                      </TableBody>
                    </Table>
                  </Grid>
                  {invoice?.notes && (
                    <div className={classes.textColor}>
                      <div className={classes.secondarySubtitle}>Notes:</div>
                      <div>{invoice.notes}</div>
                    </div>
                  )}
                  <InformationNote
                    style={{ infoContainer: { marginTop: 20 } }}
                    note={
                      <>
                        <Typography>{LANG.EN.INVOICE.PAYMENT_POLICY}</Typography>
                        <Typography>{LANG.EN.INVOICE.TAXES_CALCULTAED}</Typography>
                      </>
                    }
                  />
                  <div className={invoiceStyles.invoiceTotalsContainer}>
                    <Typography className={invoiceStyles.subTotView} align="right">
                      Subtotal: $ {formatNumberWithCommas(getFormattedDecimalNumber(invoice?.subtotal))}
                    </Typography>
                    <Typography className={invoiceStyles.subTotView} align="right">
                      Convenience Fee ({invoice?.convenienceFeeRate}%): $ {formatNumberWithCommas(getFormattedDecimalNumber(invoice?.convenienceFeeAmount))}
                    </Typography>
                    <Typography
                      className={getConditionalDefaultValue(invoice?.company?.isTaxExempt, invoiceStyles.taxExemptLabel, invoiceStyles.subTotView)}
                      style={{ marginTop: 0 }}
                      align="right"
                    >
                      {getConditionalDefaultValue(
                        invoice?.company?.isTaxExempt,
                        'This client is exempt from paying taxes.',
                        `Taxes (${invoice?.taxRate}%): $ ${formatNumberWithCommas(getFormattedDecimalNumber(invoice?.taxAmount))}`
                      )}
                    </Typography>
                    <Typography className={invoiceStyles.totalLine} align="right" component="h1" variant="h6">
                      Total: $ {formatNumberWithCommas(getFormattedDecimalNumber(invoice?.total))}
                    </Typography>
                  </div>
                </Card>
              </DialogContent>
            </>
          )}
        </>
      )}
    />
  );
};

export default memo(InvoiceInformation);
