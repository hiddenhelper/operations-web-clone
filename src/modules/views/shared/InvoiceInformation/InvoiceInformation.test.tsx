import React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createMockStore } from 'redux-test-utils';
import { MemoryRouter } from 'react-router';
import { getInvoice_1 } from '../../../../test/entities';
import { getInitialState } from '../../../../test/rootState';

import InvoiceInformation, { IInvoiceInformationProps } from './InvoiceInformation';

describe('InvoiceList', () => {
  let props: IInvoiceInformationProps;

  beforeEach(() => {
    props = {
      invoiceId: getInvoice_1().id,
      invoiceMap: {
        [getInvoice_1().id]: getInvoice_1(),
      },
      loading: undefined,
      downloadLoading: undefined,
      onClose: jest.fn(),
      fetchInvoiceInformation: jest.fn(),
      downloadInvoice: jest.fn(),
    };
  });

  it('should render', () => {
    const { container } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <InvoiceInformation {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should render tax exempt', () => {
    props.invoiceMap = {
      [getInvoice_1().id]: { ...getInvoice_1(), company: { id: '1', name: 'test', isTaxExempt: true } },
    };
    const { container } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <InvoiceInformation {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should show loading', () => {
    props.loading = {
      isLoading: true,
      hasError: false,
      error: null,
    };
    const { container } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <InvoiceInformation {...props} />
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should download', () => {
    const { getByTestId } = render(
      <Provider store={createMockStore(getInitialState()) as any}>
        <MemoryRouter>
          <InvoiceInformation {...props} />
        </MemoryRouter>
      </Provider>
    );

    act(() => {
      fireEvent.click(getByTestId('download-invoice'));
    });

    expect(props.downloadInvoice).toHaveBeenCalled();
  });
});
