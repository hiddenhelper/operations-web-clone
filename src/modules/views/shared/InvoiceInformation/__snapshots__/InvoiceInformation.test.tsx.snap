// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceList should render 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`InvoiceList should render tax exempt 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`InvoiceList should show loading 1`] = `
<div
  aria-hidden="true"
/>
`;
